pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -741,6 +751,39 @@ func (c *Controller) AppendInstanceHandler(f func(*model.ServiceInstance, model.
 		}
 		return nil
 	})
+
+	c.services.handler.Append(func(obj interface{}, event model.Event) error {
+		svc, ok := obj.(*v1.Service)
+		if !ok {
+			return nil
+		}
+
+		if svc.Namespace == meta_v1.NamespaceSystem || svc.Spec.Type != v1.ServiceTypeExternalName {
+			return nil
+		}
+
+		svcConv := convertService(*svc, c.domainSuffix)
+		instances := externalNameServiceInstances(*svc, svcConv)
+		switch event {
+		case model.EventDelete:
+			c.Lock()
+			delete(c.externalNameSvcInstanceMap, svcConv.Hostname)
+			c.Unlock()
+		default:
+			c.Lock()
+			if instances != nil {
+				c.externalNameSvcInstanceMap[svcConv.Hostname] = instances
```

Comment:
But I have another concern: How can externalName service instances update when the service object stays unchanged? It is an issue deserves consideration.

Reply:
If `AppendServiceHandler()` is called in the same context as `AppendInstanceHandler()` (which it is), then both services and instances should receive updates at about the same time.

Reply:
There's no need! Pods don't have anything to do with ExternalName services. Updates to services do trigger these event handlers for both services and instances of this type.

Reply:
My question is there is no way to update `externalNameSvcInstanceMap[svcConv.Hostname]` when pod(add/update/delete).

Reply:
But `externalNameSvcInstanceMap` is used to get serviceInstances for ExternalName service.